Moving evars_of_term from constr to econstr

herbelin opened the pull request:
**Kind:** internal, cleanup

Overlay: mattam82/Coq-Equations#150

This is an experiment in moving `evars_of_term` from type `constr -> Evar.Set.t` to `evar_map -> econstr -> Evar.Set.t`

I don't know the history of `evars_of_term` and I don't know whether there are good reasons that it is on `constr` and not on `econstr`. In any case, moving it to `econstr` allows to get rid of a number of coercions back and forth from `constr` to `econstr`. So this suggests that this move is "natural".

I was a bit unsure around `obligations.ml` and `comProgramFixpoint.ml`. I started from the assumption that `Obligations.eterm_obligations` and `Obligations.subst_evar_constr` were taking a term with evars and were returning a term where evars where replaced by constants, so morally functions from `econstr` to `constr`. A confirmation would be good.

I'm a bit unsure about `Proof_global.start_proof`: it takes an `econstr` but does not seem to be able to eventually "ground" the undefined evars of its input. A priori, this is sound to start a proof with unresolved evars, but something remains to be done for it to be really the case, and in the meantime, the API is misguiding.

In ssr, I used the sigma nearby which seemed relevant but I do not guarantee that I did not make a mistake.

This is on top of #8892.

mattam82 commented:
I confirm eterm_obligations moves from an econstr to a constr

herbelin commented:
The errors seemed to have been be spurious (no error when running ci locally). Rebased and double-checking.

ejgallego commented:
> The errors seemed to have been be spurious (no error when running ci locally). Rebased and double-checking.

Yup, they are due to some "overlays" not having been merged. You can just click "Rebuild" in the particular job once the overlays are in if you want to get the green status

ppedrot commented:
@herbelin You should rebase the overlay for Equations, it is outdated it seems.

vbgl commented:
Rebased. Overlays are probably out-of-date.

herbelin commented:
> Overlays are probably out-of-date.

I updated the Equations overlay. I could not reproduce locally the `fcsl-pcm` failure. Also, `make validate` seems locally ok with the branch.

Thanks a lot for the rebasing. I fixed the new conflict with #9861 so that it re-runs ci.

vbgl commented:
The overlay for equations seems outdated again…

herbelin commented:
> The overlay for equations seems outdated again…

The overlay was not taken into account due to the `Equations` -> `equations` renaming.

Rerunning.

vbgl commented:
@ppedrot If this is ready, that should be merged. Thanks.

vbgl commented:
The overlay for Equations is outdated.

The job `validate:quick` fails with the following message:

> Fatal Error: Anomaly "Uncaught exception Reduction.NotConvertible."

ppedrot commented:
@herbelin can you rebase this PR and its associated overlays? If CI passes I'll merge.

herbelin commented:
`make validate` is still failing on ci but I still can't reproduce locally. The 500 files I get in `theories` and `plugins` by `find . -name \*.vo` pass `coqchk`. Any hint on how to diagnose what happens on gitlab with this test?

SkySkimmer commented:
You can download the `build:quick` artifact https://gitlab.com/coq/coq/-/jobs/206921796/artifacts/download then run `bin/coqchk -o -m -coqlib lib/coq/ -norec Coq.Classes.CEquivalence` from its `_install_ci` directory to reproduce.

herbelin commented:
OK, thanks a lot. I'll try that when I have a Linux machine at hand. (If you already succeeded to get a backtrace, or more details, I'd otherwise be happy to get them.)

herbelin commented:
@skyskimmer: I could run the artifact and obtained the anomaly but I still cannot reproduce. It fails on type-checking `Coq.Classes.CEquivalence.pointwise_equivalence` and, indeed, the type in the artifact is mentioning universes about which constraints are missing. When I run locally, I have a different type for which no constraints is missing. More precisely, in the artifact, `Print pointwise_equivalence` gives:
```
fun (A : Type@{pointwise_reflexive.u2})  ...
     : forall (A : Type@{pointwise_equivalence.u2}) ...
```
with a missing constraint `pointwise_reflexive.u2 = pointwise_equivalence.u2`, while locally, I get a more convincing
```
fun (A : Type@{pointwise_reflexive.u2})  ...
     : forall (A : Type@{pointwise_reflexive.u2}) ...
```
In any case, the kernel should not have accepted the version present in the artifact. What are the exact compilation flags used in the artifact, so that I can hopefully reproduce?

For the anomaly, I opened PR #10068 so that it is now a regular typing error.

Otherwise, there is an extra strange thing in the artifact but not on my local copy of the PR: `Print pointwise_equivalence` reports universe constraints of the form `pointwise_reflexive.u0 = pointwise_reflexive.u0` which are trivial. Is that normal?

SkySkimmer commented:
>What are the exact compilation flags used in the artifact, so that I can hopefully reproduce?

It's running `make QUICK=1` so `coqc -quick` followed by `coqc -schedule-vio2vo`

herbelin commented:
@skyskimmer: thanks, I could reproduce doing
```
bin/coqc -quick theories/Classes/CEquivalence.v
bin/coqc -vio2vo theories/Classes/CEquivalence
bin/coqchk Coq.Classes.CEquivalence
```
The issue is related to the change `EConstr.Unsafe.to_constr t` into `Proof.in_proof proof (fun m -> EConstr.to_constr m t)` made in `Proof_global.close_proof` by this PR, something which is suggested to be incorrect in `allow_deferred` mode, if I understand the commit message of bf0499bc5 well enough.

I'm not familiar with this part of the code, and I don't understand well the expected invariants (*). I feel that understanding the invariants and understanding why the removal of `EConstr.Unsafe.to_constr` breaks the invariants would take long. So, I withdrew the removal of `EConstr.Unsafe.to_constr` in this PR.

(*) Another [comment](https://github.com/coq/coq/blob/45dcc1248cd2bdf00a2bcead69d6b2ed78afc51d/proofs/proof_global.ml#L206-L208) says that normalization (of evars) is needed. So, suspectfully, the current master version (i.e. no `nf` [here](https://github.com/coq/coq/blob/7fd23f30ef6c52aabc8d1084171d906e1080f3f2/proofs/proof_global.ml#L222) in the deferred case) is presumably wrong in the deferred case when there are dependencies between initial goals, isn't it?

With respect to the presence of trivial equalities in the graph of universes I mentioned above, it seems to come from function `Acyclicgraph.constraints_for` which adds equations of the form `(u,Eq,v)` w/o checking that `v` is different from `u` (assuming that the names shown by the printer are in correspondence with the internal number). But I observe these trivial equations only in the deferred case together with the `Proof.in_proof proof (fun m -> EConstr.to_constr m t)` made by this PR. So, maybe it is just again that the PR breaks an invariant.

SkySkimmer commented on dev/ci/user-overlays/08893-herbelin-master+moving-evars-of-term-on-econstr.sh:
```diff
@@ -0,0 +1,11 @@
+    equations_CI_REF=master+fix-evars_of_term-pr8893
+    equations_CI_GITURL=https://github.com/herbelin/Coq-Equations
+
+fi
```
Please list the overlay PR in the Coq PR's first message
Also you only need the 2nd triple (there is no `Equations_CI_FOO` variable)

herbelin replied:
Done

ejgallego commented:
@herbelin I think you pushed some commits not intended for this branch.

herbelin commented:
Hum, very sorry. I should have double-checked. Now really pushed the overlay change and moved the wrong commits to the branch they pertain to.

ppedrot commented:
I have trouble following this PR. Am I right that this is ready modulo CI?

herbelin commented:
> Am I right that this is ready modulo CI?

I think so. CI was green before it was restarted after I remove the useless lines of the overlay.

Note that I backtracked on the change I had made in `Proof.close_proof` which were causing a failure of `coqchk`. (I did not explore it further, but it seems that `Proof.close_proof` is able to make the kernel validate proofs missing universe constraints, since, by replacing the `EConstr.Unsafe.to_constr` by one using the sigma, one can miss constraints between the body and type of a deferred theorem...)

ppedrot commented:
CI failure spurious, I rebased the overlay and checked locally.

hendriktews commented:
Commit 6608f64f001f8f1a50b2dc41fefdf63c0b84b270 in this PR causes the dependent evars line to always be empty, see also #10399. Was that change intentional?
